File: packages/sdkwork-utils-typescript/src/i18n.ts
import { round } from "./number.js";
import { parseDatetime } from "./datetime.js";

function separators(locale: string): { decimal: string; grouping: string } {
  if (locale.toLowerCase() === "de-de") {
    return { decimal: ",", grouping: "." };
  }
  return { decimal: ".", grouping: "," };
}

export function formatNumberLocale(value: number, locale: string, decimals = 0): string {
  const rounded = round(Math.abs(value), decimals);
  const { decimal, grouping } = separators(locale);
  const [integerPart, fractionPart = ""] = rounded.toFixed(decimals).split(".");
  const grouped = integerPart.replace(/\B(?=(\d{3})+(?!\d))/g, grouping);
  const formatted =
    decimals > 0 ? `${grouped}${decimal}${fractionPart.padEnd(decimals, "0")}` : grouped;
  return value < 0 ? `-${formatted}` : formatted;
}

export function formatDatetimeLocale(value: Date, locale: string): string {
  const normalized = locale.toLowerCase();
  const year = value.getUTCFullYear();
  const month = String(value.getUTCMonth() + 1).padStart(2, "0");
  const day = String(value.getUTCDate()).padStart(2, "0");
  const hour = String(value.getUTCHours()).padStart(2, "0");
  const minute = String(value.getUTCMinutes()).padStart(2, "0");
  if (normalized.startsWith("de")) {
    return `${day}.${month}.${year} ${hour}:${minute}`;
  }
  if (normalized.startsWith("zh")) {
    return `${year}-${month}-${day} ${hour}:${minute}`;
  }
  return `${month}/${day}/${year} ${hour}:${minute}`;
}

export function formatDatetimeLocaleStr(value: string, locale: string): string | null {
  const parsed = parseDatetime(value);
  return parsed ? formatDatetimeLocale(parsed, locale) : null;
}

export function parseNumberLocale(input: string, locale: string): number | null {
  const trimmed = input.trim();
  if (!trimmed) {
    return null;
  }
  const { decimal, grouping } = separators(locale);
  const normalized = trimmed.replaceAll(grouping, "").replace(decimal, ".");
  const parsed = Number.parseFloat(normalized);
  return Number.isFinite(parsed) ? parsed : null;
}

File: packages/sdkwork-utils-python/sdkwork_utils/i18n.py
from datetime import datetime

from sdkwork_utils.datetime import parse_datetime
from sdkwork_utils.number import round_number


def _separators(locale: str) -> tuple[str, str]:
    if locale.lower() == "de-de":
        return ",", "."
    return ".", ","


def format_number_locale(value: float, locale: str, decimals: int = 0) -> str:
    rounded = round_number(abs(value), decimals)
    decimal_sep, grouping_sep = _separators(locale)
    integer_part, _, fraction_part = f"{rounded:.{decimals}f}".partition(".")
    grouped = ""
    for index, digit in enumerate(reversed(integer_part)):
        if index > 0 and index % 3 == 0:
            grouped = grouping_sep + grouped
        grouped = digit + grouped
    if decimals > 0:
        formatted = f"{grouped}{decimal_sep}{fraction_part.ljust(decimals, '0')}"
    else:
        formatted = grouped
    return f"-{formatted}" if value < 0 else formatted


def format_datetime_locale(value: datetime, locale: str) -> str:
    normalized = locale.lower()
    if normalized.startswith("de"):
        return value.strftime("%d.%m.%Y %H:%M")
    if normalized.startswith("zh"):
        return value.strftime("%Y-%m-%d %H:%M")
    return value.strftime("%m/%d/%Y %H:%M")


def format_datetime_locale_str(value: str, locale: str) -> str | None:
    parsed = parse_datetime(value)
    return format_datetime_locale(parsed, locale) if parsed is not None else None


def parse_number_locale(input: str, locale: str) -> float | None:
    trimmed = input.strip()
    if not trimmed:
        return None
    decimal_sep, grouping_sep = _separators(locale)
    normalized = trimmed.replace(grouping_sep, "").replace(decimal_sep, ".")
    try:
        return float(normalized)
    except ValueError:
        return None
